### ISSUE TITLE
Undo feature may have undesired results

### ISSUE THREAD

**Issue by WinnerEX** [processing/processing#4775]
Undo within the processing environment can create problems and has created issues for me. Under heavy cpu load where the screen might not update in real time, holding ctrl+z can lead to erratic behaviour of Undo that can only best be described as it having forgotten the location of the cursor during its history. Random operators or line breaks may be undone leading to a mess of code.

Unsure of how to reliably reproduce, but perhaps this next issue may be a part of the fault:

Double-clicking or highlighting a portion of code and replacing it with a paste from ctrl+v, then Undoing that action will instead altogether delete whatever was there instead of replacing it. Example
<code>int var = 60;</code>
Highlighting 60, ctrl+v "139"
<code>int var = 139</code>
ctrl+z
<code>int var = </code>

Expected behaviour is return to previous, "60".
A second iteration of Undo does restore the previous overwritten value but needing to undo twice for a single action is not expected.

Alternatively, pasting multiple times successively is also completely undone by a single undo

<code>int var = 60;</code>
Highlighting 60, ctrl+v "139"
<code>int var = 139</code>
paste paste paste paste paste paste...
<code>int var = 139139139139139139139</code>
ctrl+z
<code>int var = </code>

I also just now had an issue where after pasting something, my cursor jumped ahead about 8 characters and it was halfway into the name of a variable following what i'd just pasted. Unable to reproduce at this moment but will update if i figure out what it was.

**Comment by WinnerEX** [comment]
Update: the issue with the jumping cursor is related to replacing a highlighted selection with newly input info, then undoing.
Example
<code>centerx-v, centery+v</code>
Double-click highlight centerx, replace "1"
<code>1-v, centery+v</code>
cursor is now after "1"
ctrl+z
<code>centerx-v, centery+v</code>
cursor is now between "n" and "t" in "centery"

**Comment by benfry** [comment]
Are you using the latest release?

**Comment by jeremydouglass** [comment]
I've experienced this in the past with 3.x., and recently seen people in the forums report it yet again https://discourse.processing.org/t/undo-and-redo-errors-pde-ide/16419/4

The key might be using undo multiple times in quick succession -- holding Ctrl-Z down (perhaps accidentally with a long tap), or rapid-tapping it 2-3 times. Perhaps there is a race condition where the first undo operation hasn't finished updating when the second begins.

> Still, it is hard to replicate – I just did a quick test and was unable to replicate the error on 3.4, even while running a sketch under heavy load – holding Ctr+Z and Ctrl-Shift-Z scrubbed back and forth through a long edit history with no errors.

**Comment by DaveJarvis** [comment]
``` plain
javax.swing.text.BadLocationException: Length must be positive
	at javax.swing.text.AbstractDocument.getText(AbstractDocument.java:808)
	at processing.app.syntax.JEditTextArea.getText(JEditTextArea.java:1033)
	at processing.app.syntax.JEditTextArea.getLineText(JEditTextArea.java:1062)
	at processing.app.syntax.TextAreaPainter.paintSyntaxLine(TextAreaPainter.java:663)
	at processing.app.syntax.TextAreaPainter.paintLine(TextAreaPainter.java:617)
	at processing.app.syntax.PdeTextAreaPainter.paintLine(PdeTextAreaPainter.java:120)
	at processing.app.syntax.TextAreaPainter.paint(TextAreaPainter.java:495)
	at processing.mode.java.pdex.JavaTextAreaPainter.paint(JavaTextAreaPainter.java:91)
	at javax.swing.JComponent.paintToOffscreen(JComponent.java:5210)
	at javax.swing.BufferStrategyPaintManager.paint(BufferStrategyPaintManager.java:290)
	at javax.swing.RepaintManager.paint(RepaintManager.java:1272)
	at javax.swing.JComponent._paintImmediately(JComponent.java:5158)
	at javax.swing.JComponent.paintImmediately(JComponent.java:4969)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:831)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:814)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:814)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:789)
	at javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:738)
	at javax.swing.RepaintManager.access$1200(RepaintManager.java:64)
	at javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1732)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Bad Location: -3774 for start 4826 and length -3774javax.swing.text.BadLocationException: Length must be positive
	at javax.swing.text.AbstractDocument.getText(AbstractDocument.java:808)
	at processing.app.syntax.JEditTextArea.getText(JEditTextArea.java:1033)
	at processing.app.syntax.JEditTextArea.getLineText(JEditTextArea.java:1062)
	at processing.app.syntax.JEditTextArea.xToOffset(JEditTextArea.java:690)
	at processing.app.syntax.PdeTextArea.xToOffset(PdeTextArea.java:154)
	at processing.app.syntax.InputHandler$prev_line.actionPerformed(InputHandler.java:965)
	at processing.app.syntax.InputHandler.executeAction(InputHandler.java:308)
	at processing.app.syntax.DefaultInputHandler.keyPressed(DefaultInputHandler.java:219)
	at processing.app.syntax.PdeInputHandler.keyPressed(PdeInputHandler.java:245)
	at processing.app.syntax.JEditTextArea.processKeyEvent(JEditTextArea.java:1891)
	at processing.mode.java.pdex.JavaTextArea.processKeyEvent(JavaTextArea.java:145)
	at java.awt.Component.processEvent(Component.java:6316)
	at java.awt.Container.processEvent(Container.java:2239)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2297)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1954)
	at java.awt.DefaultKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:835)
	at java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1103)
	at java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:974)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:800)
	at java.awt.Component.dispatchEventImpl(Component.java:4760)
	at java.awt.Container.dispatchEventImpl(Container.java:2297)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:760)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:84)
	at java.awt.EventQueue$4.run(EventQueue.java:733)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:730)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Bad Location: -3774 for start 4826 and length -3774javax.swing.text.BadLocationException: Length must be positive
	at javax.swing.text.AbstractDocument.getText(AbstractDocument.java:808)
	at processing.app.syntax.JEditTextArea.getText(JEditTextArea.java:1033)
	at processing.app.syntax.JEditTextArea.getLineText(JEditTextArea.java:1062)
	at processing.app.syntax.JEditTextArea.xToOffset(JEditTextArea.java:690)
	at processing.app.syntax.PdeTextArea.xToOffset(PdeTextArea.java:154)
	at processing.app.syntax.InputHandler$prev_line.actionPerformed(InputHandler.java:965)
	at processing.app.syntax.InputHandler.executeAction(InputHandler.java:308)
	at processing.app.syntax.DefaultInputHandler.keyPressed(DefaultInputHandler.java:219)
	at processing.app.syntax.PdeInputHandler.keyPressed(PdeInputHandler.java:245)
	at processing.app.syntax.JEditTextArea.processKeyEvent(JEditTextArea.java:1891)
	at processing.mode.java.pdex.JavaTextArea.processKeyEvent(JavaTextArea.java:145)
	at java.awt.Component.processEvent(Component.java:6316)
	at java.awt.Container.processEvent(Container.java:2239)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2297)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1954)
	at java.awt.DefaultKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:835)
	at java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1103)
	at java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:974)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:800)
	at java.awt.Component.dispatchEventImpl(Component.java:4760)
	at java.awt.Container.dispatchEventImpl(Container.java:2297)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:760)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:84)
	at java.awt.EventQueue$4.run(EventQueue.java:733)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:730)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Bad Location: -3774 for start 4826 and length -3774javax.swing.text.BadLocationException: Length must be positive
	at javax.swing.text.AbstractDocument.getText(AbstractDocument.java:808)
	at processing.app.syntax.JEditTextArea.getText(JEditTextArea.java:1033)
	at processing.app.syntax.JEditTextArea.getLineText(JEditTextArea.java:1062)
	at processing.app.syntax.JEditTextArea.xToOffset(JEditTextArea.java:690)
	at processing.app.syntax.PdeTextArea.xToOffset(PdeTextArea.java:154)
	at processing.app.syntax.InputHandler$prev_line.actionPerformed(InputHandler.java:965)
	at processing.app.syntax.InputHandler.executeAction(InputHandler.java:308)
	at processing.app.syntax.DefaultInputHandler.keyPressed(DefaultInputHandler.java:219)
	at processing.app.syntax.PdeInputHandler.keyPressed(PdeInputHandler.java:245)
	at processing.app.syntax.JEditTextArea.processKeyEvent(JEditTextArea.java:1891)
	at processing.mode.java.pdex.JavaTextArea.processKeyEvent(JavaTextArea.java:145)
	at java.awt.Component.processEvent(Component.java:6316)
	at java.awt.Container.processEvent(Container.java:2239)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2297)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.KeyboardFocusManager.redispatchEvent(KeyboardFocusManager.java:1954)
	at java.awt.DefaultKeyboardFocusManager.dispatchKeyEvent(DefaultKeyboardFocusManager.java:835)
	at java.awt.DefaultKeyboardFocusManager.preDispatchKeyEvent(DefaultKeyboardFocusManager.java:1103)
	at java.awt.DefaultKeyboardFocusManager.typeAheadAssertions(DefaultKeyboardFocusManager.java:974)
	at java.awt.DefaultKeyboardFocusManager.dispatchEvent(DefaultKeyboardFocusManager.java:800)
	at java.awt.Component.dispatchEventImpl(Component.java:4760)
	at java.awt.Container.dispatchEventImpl(Container.java:2297)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:760)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:84)
	at java.awt.EventQueue$4.run(EventQueue.java:733)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:730)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Bad Location: -3774 for start 4826 and length -3774javax.swing.text.BadLocationException: Length must be positive
	at javax.swing.text.AbstractDocument.getText(AbstractDocument.java:808)
	at processing.app.syntax.JEditTextArea.getText(JEditTextArea.java:1033)
	at processing.app.syntax.JEditTextArea.getLineText(JEditTextArea.java:1062)
	at processing.app.syntax.TextAreaPainter.paintSyntaxLine(TextAreaPainter.java:663)
	at processing.app.syntax.TextAreaPainter.paintLine(TextAreaPainter.java:617)
	at processing.app.syntax.PdeTextAreaPainter.paintLine(PdeTextAreaPainter.java:120)
	at processing.app.syntax.TextAreaPainter.paint(TextAreaPainter.java:495)
	at processing.mode.java.pdex.JavaTextAreaPainter.paint(JavaTextAreaPainter.java:91)
	at javax.swing.JComponent.paintToOffscreen(JComponent.java:5210)
	at javax.swing.BufferStrategyPaintManager.paint(BufferStrategyPaintManager.java:290)
	at javax.swing.RepaintManager.paint(RepaintManager.java:1272)
	at javax.swing.JComponent._paintImmediately(JComponent.java:5158)
	at javax.swing.JComponent.paintImmediately(JComponent.java:4969)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:831)
	at javax.swing.RepaintManager$4.run(RepaintManager.java:814)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:814)
	at javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:789)
	at javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:738)
	at javax.swing.RepaintManager.access$1200(RepaintManager.java:64)
	at javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1732)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Bad Location: -3774 for start 4826 and length -3774javax.swing.text.BadLocationException: Length must be positive
	at javax.swing.text.AbstractDocument.getText(AbstractDocument.java:808)
	at processing.app.syntax.JEditTextArea.getText(JEditTextArea.java:1033)
	at processing.app.syntax.JEditTextArea.getLineText(JEditTextArea.java:1062)
	at processing.app.syntax.TextAreaPainter.paintSyntaxLine(TextAreaPainter.java:663)
	at processing.app.syntax.TextAreaPainter.paintLine(TextAreaPainter.java:617)
	at processing.app.syntax.PdeTextAreaPainter.paintLine(PdeTextAreaPainter.java:120)
	at processing.app.syntax.TextAreaPainter.paint(TextAreaPainter.java:495)
	at processing.mode.java.pdex.JavaTextAreaPainter.paint(JavaTextAreaPainter.java:91)
	at javax.swing.JComponent.paintToOffscreen(JComponent.java:5210)
	at javax.swing.BufferStrategyPaintManager.paint(BufferStrategyPaintManager.java:290)
```

**Comment by dzaima** [comment]
I can reproduce issues with undoing when using multiple tabs:

1. write `foo` in the first tab
2. write `bar` _quickly after_ in the second tab
3. undo
4. typing in the first tab now errors

It seems Processing isn't properly handling the case when a single undo spans multiple files (though it doesn't have problems when the undo is from a rename).

**Comment by DaveJarvis** [comment]
> It seems Processing isn't properly handling the case when a single undo spans multiple files (though it doesn't have problems when the undo is from a rename).

Ideally, each tab would have its own "undo" and "redo" history.

**Comment by benfry** [comment]
> Ideally, each tab would have its own "undo" and "redo" history.

Of course; they already do. There's just something else broken in there, but it may be fixed in 4.0 alpha 4 with https://github.com/processing/processing4/pull/175

**Comment by hkiel** [comment]
Can you backport the fix to processing3?

**Comment by benfry** [comment]
No, I barely have time to maintain Processing 4, I can't maintain both 3 and 4.

**Comment by StanLepunK** [comment]
May be it's time to close the repository of Processing 3 or it's to soon ?

**Comment by benfry** [comment]
Too soon; we don't have a final release of 4, and it's not the default download.

**Comment by benfry** [comment]
Should be fixed in the 4.0 releases: https://github.com/processing/processing4/releases
